File: app/src/main/java/com/linheimx/zimudog/vp/search/P_Zimuku.java
package com.linheimx.zimudog.vp.search;

import android.support.annotation.NonNull;

import com.linheimx.lspider.god.GodPage;
import com.linheimx.zimudog.m.net.ApiManager;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;

/**
 * Created by x1c on 2017/5/1.
 */

public class P_Zimuku extends P_Base implements IContract.P {

    public P_Zimuku(IContract.V v) {
        super(v);
    }

    String _movie;
    int _page = 1;

    @Override
    public void searchByKW(@NonNull String keyWord) {
        this._movie = keyWord;
        this._page = 1;
        loadMovies(_movie, _page);
    }

    @Override
    public void loadMore() {
        loadMovies(_movie, ++_page);
    }

    private void loadMovies(String movie, int page) {

        ApiManager.getInstence()
                .getPage_Zimuku(movie, page)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<GodPage>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        _Disposables.add(d);
                    }

                    @Override
                    public void onNext(GodPage page) {
                        _V.showItems(page.getItemList(), page.isHasMore());
                    }

                    @Override
                    public void onError(Throwable e) {
                        _V.showLoadingError();
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: app/src/main/java/com/linheimx/zimudog/vp/search/P_Zimuku.kt
package com.linheimx.zimudog.vp.search

import android.util.Log
import com.google.gson.Gson
import com.linheimx.zimudog.m.bean.Resp_Movies
import com.linheimx.zimudog.m.net.ApiManger
import com.linheimx.zimudog.vp.base.BasePresenter
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

/**
 * Created by x1c on 2017/5/1.
 */

class P_Zimuku(var v: IContract.V) : BasePresenter(), IContract.P {

    var _movie: String = ""
    var _page = 1

    override fun cancelSearch() {
        _Disposables.dispose()
        _Disposables = CompositeDisposable()
    }

    override fun release() {
        _Disposables.dispose()
    }

    override fun searchByKW(keyWord: String) {
        this._movie = keyWord
        this._page = 1
        loadMovies(_movie, _page)
    }

    override fun loadMore() {
        loadMovies(_movie, ++_page)
    }

    private fun loadMovies(movie: String, page: Int) {

        ApiManger.api.getHtml_MovieList(movie, page)
                .subscribeOn(Schedulers.io())
                .flatMap {
                    ApiManger.apiParse.parse_MovieList(it.string())
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<Resp_Movies> {
                    override fun onComplete() {
                    }

                    override fun onSubscribe(d: Disposable) {
                        _Disposables.add(d)
                    }

                    override fun onNext(it: Resp_Movies) {
                        Log.e("--->", "ret:" + Gson().toJson(it))
                        var page = it.obj
                        if (page == null) {
                            v.showNoDataView()
                        } else {
                            v.showItems(page.movies, page.haveNext)
                        }
                    }

                    override fun onError(it: Throwable) {
                        Log.e("--->", "load error:" + it)
                        v.showLoadingError(it)
                    }
                })
    }
}
